### ISSUE TITLE
Cannot quantize part of a model

### ISSUE THREAD

**Issue by xumengwei** [tensorflow/tensorflow#46073]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.5.0-dev20201208
- Python version: 3.7

**Standalone code to reproduce the issue**
I follow the tutorial [here](https://www.tensorflow.org/model_optimization/guide/quantization/training_comprehensive_guide) to see if I can only quantize part of a model. Here is my code:

```
i = tf.keras.Input(shape=(20,))
x = tfmot.quantization.keras.quantize_annotate_layer(tf.keras.layers.Dense(10))(i)
x = tf.keras.layers.Dense(10)(x)
x = tfmot.quantization.keras.quantize_annotate_layer(tf.keras.layers.Dense(10))(x)
o = tf.keras.layers.Flatten()(x)
annotated_model = tf.keras.Model(inputs=i, outputs=o)
quant_aware_model = tfmot.quantization.keras.quantize_apply(annotated_model)
converter = tf.lite.TFLiteConverter.from_keras_model(quant_aware_model)
converter.optimizations = [tf.lite.Optimize.DEFAULT]
quantized_tflite_model = converter.convert()
pathlib.Path('/tmp/tmp.tflite').write_bytes(quantized_tflite_model)
```
Basically I want to quantize the 1st and 3rd dense layers of a model. And here is how the resultant model looks like:
<img width="237" alt="Screen Shot 2020-12-30 at 8 02 19 PM" src="https://user-images.githubusercontent.com/23546158/103350164-0741a280-4ada-11eb-9c0e-713bb2ceddca.png">
Apparently, the 2nd quantization goes to the wrong place...

**Comment by xumengwei** [comment]
@daverim @teijeong can you help this since you are assigned to this thread? I think it's a very basic function but seems doesn't work...

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with TF v2.4 and TF-nightly. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/edfc37a8c5d517725e198274deed71e6/46073.ipynb). 
![tmp tflite](https://user-images.githubusercontent.com/57165142/103513806-5ea49180-4e91-11eb-9e29-2c4f693fb066.png)

Thanks!

**Comment by xumengwei** [comment]
@amahendrakar thanks for confirming. While you are working on this, can you also elaborate which part of the codebase can be the reason? Maybe the toco converter? I'd also want to take a look at this issue.

**Comment by daverim** [comment]
Something in the converter, definitely --- not sure why, still investigating... will update soon

**Comment by xumengwei** [comment]
@daverim Thanks! I went through the quantization code (python part), and have a feeling that the [tutorial](https://www.tensorflow.org/model_optimization/guide/quantization/training_comprehensive_guide) about "quantizing some layers" is for weights-only quantization, while I'm looking for quantizing the whole layer so it can be delegated to DSP. Is that right?
In addition, where I can find any document about how toco does graph transformation, e.g., remove fake quant op, adding quant/dequant ops, etc. I feel there's some disconnection between the high-level quant APIs and toco's model conversion logics.

**Comment by daverim** [comment]
hi @xumengwei,

It turns out that there is kind of an issue in QAT. The input (other than an Input layer) to a quantized annotated layer must also be quantized annotated in order to be fully quantized -- this is because we need to record the min/max ranges of the input to the layer in order to quantize the layer. 

Because you have the output of a non-quantized dense layer as input to a quantized layer, you end up with two float dense layers. If you add another layer the next layer will be quantized, as the input is quantized

```
import tensorflow as tf

def model():
  i = tf.keras.Input(shape=(20,))
  quantized = tfmot.quantization.keras.quantize_annotate_layer(tf.keras.layers.Dense(10, activation='relu'))(i)
  non_quantized = tf.keras.layers.Dense(10, activation='relu')(quantized)
  also_non_quantized = tfmot.quantization.keras.quantize_annotate_layer(tf.keras.layers.Dense(10, activation='relu'))(non_quantized)
  quantized_again = tfmot.quantization.keras.quantize_annotate_layer(tf.keras.layers.Dense(10, activation='relu'))(also_non_quantized)
  o = tf.keras.layers.Flatten()(quantized_again)
  model = tf.keras.Model(inputs=i, outputs=o)
  quantize_aware_model = tfmot.quantization.keras.quantize_apply(model)
  return quantize_aware_model

quantize_aware_model = model()
quantize_aware_model.summary()

converter = tf.lite.TFLiteConverter.from_keras_model(quantized)
converter.optimizations = [tf.lite.Optimize.DEFAULT]
quantized_tflite_model = converter.convert()
open('tmp.tflite', 'wb').write(bytearray(quantized_tflite_model))
```


![image](https://user-images.githubusercontent.com/2448933/103981318-f02f3e80-51c4-11eb-884b-4cc409430224.png)

This is a bug that I am working on addressing in QAT, and not really a tflite error, as the recording ops (fakequant) is not added to input tensors when expected -- in the meantime, you can use this information to change the structure of your model just be aware that for a layer to be quantized its input must also be quantized. In your case, you would need to add an additional quantized op before the layer you want to have quantized.

As for conversion logic, the new logic is basically implemented as tensorflow graphdef -> tensorflow mlir -> tflite mlir -> tflite flatbuffer. For more information you can take a look at https://www.tensorflow.org/mlir and https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/mlir/lite/README.md.

Hope that helps, will update on this later

**Comment by xumengwei** [comment]
@daverim thanks for digging into it! Regarding your revised code, I have a confusion: the layer `also_non_quantized ` is annotated as quantized; although it's not really quantized because its input is not quantized, but won't its weights be quantized anyway? If not, what's the way to do weights-only quantization for certain layers?

**Comment by ashishpatel26** [comment]
@xumengwei  Try this

```python
import tensorflow as tf
from tensorflow.keras.models import load_model
from tensorflow.keras.layers import Input
from tensorflow.keras.models import Model

def get_file_size(file_path):
    size = os.path.getsize(file_path)
    return size
    
def convert_bytes(size, unit=None):
    if unit == "KB":
        return print('File size: ' + str(round(size / 1024, 3)) + ' Kilobytes')
    elif unit == "MB":
        return print('File size: ' + str(round(size / (1024 * 1024), 3)) + ' Megabytes')
    else:
        return print('File size: ' + str(size) + ' bytes')

def convert_model_to_tflite(model_path = "model.h5", filename = f"model.tflite", modelname = "model"):
  model = model_load(model_name=modelname, weights=model_path)
  fixed_input = Input((230,230,3))
  fixed_model = Model(fixed_input,model(fixed_input))
  converter = tf.lite.TFLiteConverter.from_keras_model(model)
  converter.optimizations = [tf.lite.Optimize.OPTIMIZE_FOR_SIZE]
  converter.target_spec.supported_ops = [
    tf.lite.OpsSet.TFLITE_BUILTINS, # enable TensorFlow Lite ops.
    tf.lite.OpsSet.SELECT_TF_OPS,  # enable TensorFlow ops.
    tf.float16
  ]
  tflite_model = converter.convert()
  open(filename, "wb").write(tflite_model)
  print(convert_bytes(get_file_size(f"converted_model_Inceptionresnetv2.tflite"), "MB"))
```

**Comment by Chalay** [comment]
Yes, quantize part of the model is very important to some kind of net like landmark detection

**Comment by macoskey** [comment]
It seems to me that the issue of being able to manually select which weights to quantize when using the `TFLiteConverter` is still not resolved. There seems to be a disconnect between the tenforflow_model_optimization team, which says you can [quantize some layers (i.e., pick-and-choose which ones to quantize)](https://www.tensorflow.org/model_optimization/guide/quantization/training_comprehensive_guide#quantize_some_layers), and the tflite team, which doesn't mention this capability anywhere. **Is it possible to only quantize the weights of, e.g., 4 out of 5 layers in a tflite model?**

**Comment by daverim** [comment]
It is possible to quantize per-layer using quantized aware training as you mention. You can use the quantize_annotate_layer as mentioned for during training.

For post-training quantization, you can use the quantization debugger: https://www.tensorflow.org/lite/performance/quantization_debugger#selective_quantization

Please leave feedback if you are interested in this.

**Comment by macoskey** [comment]
Thanks for your reply @daverim! I'm aware of the per-layer quantization available during training via `quantize_annotate_layer`, but what I'm after is a way to only quantize certain weights of a tflite model. For example, if I have a model with 5 LSTM layers, I'd like to have the first layer run entirely at float32 but the weights for layers 2-5 to be quantized to int8. Is this possible? Whether or not the activations and inputs/outputs are quantized is a separate topic--for now I'm just focused on weight & bias quantization.

**Comment by mohantym** [comment]
Reopening this issue .

**Comment by macoskey** [comment]
Update: It seems like you can perform [selective quantization](https://www.tensorflow.org/lite/performance/quantization_debugger#selective_quantization) with the `QuantizationDebugger`.  However, you have to know the exact node(s) you'd like to leave at float32, which isn't a trivial task because the nodes don't have names that clearly map them back to Keras layers. I'm also finding that the `QuantizationDebugger` is flaky with custom architectures.

Regardless, this capability isn't built into the `TFLiteConverter`. It seems that one could build a wrapper around `tensorflow.lite.python.convert.mlir_quantize`, but this feels like a missing feature of the TFLite library given all the emphasis on selective quantization both in the debugger and in the QAT fake quant nodes.

**Comment by DerryFitz** [comment]
Any update on adding selective quantization of layers to TfLiteConverter, I'd like to see this added too

**Comment by tgsmith61591** [comment]
Would be very interested in this as well..

**Comment by pkgoogle** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The TFLite team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the TFLite space.

Thanks.
